File: sdk/rust/src/canonical.rs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
//! Canonical JSON serialization and context identity (§10).

use crate::types::HookContext;
use serde_json::Value;
use sha2::{Digest, Sha256};
use std::fmt::Write;

const L0: &[&str] = &[
    "spec",
    "hook_point",
    "timestamp",
    "sequence",
    "agent",
    "session",
    "target",
];
const L0_AGENT: &[&str] = &["id", "framework"];
const L0_SESSION: &[&str] = &["id"];

fn l1_for(hp: &str) -> &'static [&'static str] {
    match hp {
        "agent_startup" => &["agent_init"],
        "input" => &["input"],
        "pre_model_call" => &["model", "messages"],
        "post_model_call" => &["model", "response"],
        "pre_tool_call" => &["tool_call"],
        "post_tool_call" => &["tool_call", "tool_result"],
        "output" => &["output"],
        "agent_shutdown" => &["summary"],
        _ => &[],
    }
}

fn encode(v: &Value, out: &mut String) {
    match v {
        Value::Null => out.push_str("null"),
        Value::Bool(b) => out.push_str(if *b { "true" } else { "false" }),
        Value::Number(n) => {
            // serde_json::Number already produces shortest-round-trip;
            // strip trailing .0 to match ECMA-262 / cross-SDK identities.
            let s = n.to_string();
            out.push_str(s.strip_suffix(".0").unwrap_or(&s));
        }
        Value::String(s) => {
            // RFC 8259 minimal escapes via serde_json's serializer.
            out.push_str(&serde_json::to_string(s).expect("string serialize"));
        }
        Value::Array(a) => {
            out.push('[');
            for (i, e) in a.iter().enumerate() {
                if i > 0 {
                    out.push(',');
                }
                encode(e, out);
            }
            out.push(']');
        }
        Value::Object(m) => {
            out.push('{');
            let mut keys: Vec<_> = m.keys().collect();
            keys.sort_by(|a, b| a.as_bytes().cmp(b.as_bytes()));
            for (i, k) in keys.iter().enumerate() {
                if i > 0 {
                    out.push(',');
                }
                out.push_str(&serde_json::to_string(k).expect("key serialize"));
                out.push(':');
                encode(&m[*k], out);
            }
            out.push('}');
        }
    }
}

/// Serialize `v` per §10.1: lexicographic keys, no whitespace, ECMA-262
/// numbers, RFC 8259 minimal string escapes.
pub fn canonical_json(v: &Value) -> String {
    let mut out = String::new();
    encode(v, &mut out);
    out
}

fn strip_to_l01(ctx: &HookContext) -> Value {
    let hp = ctx
        .get("hook_point")
        .and_then(Value::as_str)
        .unwrap_or_default();
    let l1 = l1_for(hp);
    let mut out = serde_json::Map::new();
    for (k, v) in ctx {
        if !L0.contains(&k.as_str()) && !l1.contains(&k.as_str()) {
            continue;
        }
        let v = match k.as_str() {
            "agent" => filter_obj(v, L0_AGENT),
            "session" => filter_obj(v, L0_SESSION),
            _ => v.clone(),
        };
        out.insert(k.clone(), v);
    }
    Value::Object(out)
}

fn filter_obj(v: &Value, keep: &[&str]) -> Value {
    match v {
        Value::Object(m) => Value::Object(
            m.iter()
                .filter(|(k, _)| keep.contains(&k.as_str()))
                .map(|(k, v)| (k.clone(), v.clone()))
                .collect(),
        ),
        other => other.clone(),
    }
}

/// `"sha256:" + hex(SHA-256(canonical_json(ctx_L01)))` (§10.2).
pub fn context_identity(ctx: &HookContext) -> String {
    let json = canonical_json(&strip_to_l01(ctx));
    let digest = Sha256::digest(json.as_bytes());
    let mut out = String::with_capacity(7 + 64);
    out.push_str("sha256:");
    for b in digest {
        write!(out, "{b:02x}").expect("write hex");
    }
    out
}

File: sdk/python/src/agent_hooks/canonical.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.
"""Canonical JSON serialization and context identity (§10)."""
from __future__ import annotations

import hashlib
import json
import math
from typing import Any

from agent_hooks._types import InterceptionPoint

#: L0 keys (§4.1) — always retained for identity.
_L0: frozenset[str] = frozenset(
    {"spec", "interception_point", "timestamp", "sequence", "agent", "session", "target"}
)
#: L0 sub-keys retained on ``agent`` and ``session``.
_L0_AGENT: frozenset[str] = frozenset({"id", "framework"})
_L0_SESSION: frozenset[str] = frozenset({"id"})

#: L1 keys per interception point (§4.2).
_L1: dict[str, frozenset[str]] = {
    InterceptionPoint.AGENT_STARTUP.value: frozenset({"agent_init"}),
    InterceptionPoint.INPUT.value: frozenset({"input"}),
    InterceptionPoint.PRE_MODEL_CALL.value: frozenset({"model", "messages"}),
    InterceptionPoint.POST_MODEL_CALL.value: frozenset({"model", "response"}),
    InterceptionPoint.PRE_TOOL_CALL.value: frozenset({"tool_call"}),
    InterceptionPoint.POST_TOOL_CALL.value: frozenset({"tool_call", "tool_result"}),
    InterceptionPoint.OUTPUT.value: frozenset({"output"}),
    InterceptionPoint.AGENT_SHUTDOWN.value: frozenset({"summary"}),
}


def _ecma_number(x: float) -> str:
    """ECMA-262 shortest-round-trip number serialization (§10.1.3).

    Python's ``repr`` already produces the shortest round-trip for floats
    since 3.1; we strip the trailing ``.0`` for integral values to match the
    JS ``Number.prototype.toString`` output exactly so identities agree
    across SDKs.
    """
    if math.isnan(x) or math.isinf(x):
        raise ValueError("canonical JSON does not admit NaN/Infinity")
    if x == 0:
        return "0"  # collapse -0.0
    s = repr(x)
    if s.endswith(".0"):
        return s[:-2]
    return s


def _encode(obj: Any, out: list[str]) -> None:
    if obj is None:
        out.append("null")
    elif obj is True:
        out.append("true")
    elif obj is False:
        out.append("false")
    elif isinstance(obj, str):
        out.append(json.dumps(obj, ensure_ascii=False))
    elif isinstance(obj, bool):  # pragma: no cover — handled above, but guard ordering
        out.append("true" if obj else "false")
    elif isinstance(obj, int):
        out.append(str(obj))
    elif isinstance(obj, float):
        out.append(_ecma_number(obj))
    elif isinstance(obj, (list, tuple)):
        out.append("[")
        for i, v in enumerate(obj):
            if i:
                out.append(",")
            _encode(v, out)
        out.append("]")
    elif isinstance(obj, dict):
        out.append("{")
        for i, k in enumerate(sorted(obj.keys(), key=lambda s: s.encode("utf-8"))):
            if i:
                out.append(",")
            out.append(json.dumps(k, ensure_ascii=False))
            out.append(":")
            _encode(obj[k], out)
        out.append("}")
    else:
        raise TypeError(f"canonical JSON cannot encode {type(obj).__name__}")


def canonical_json(obj: Any) -> str:
    """Serialize ``obj`` per §10.1: lexicographic keys, no whitespace,
    ECMA-262 numbers, RFC 8259 minimal string escapes."""
    buf: list[str] = []
    _encode(obj, buf)
    return "".join(buf)


def _strip_to_l01(ctx: dict[str, Any]) -> dict[str, Any]:
    """Return a deep copy of ``ctx`` containing only L0 + L1 fields (§10.2)."""
    hp = ctx["interception_point"]
    keep = _L0 | _L1.get(hp, frozenset())
    out: dict[str, Any] = {}
    for k in ctx:
        if k not in keep:
            continue
        if k == "agent":
            out[k] = {sk: ctx[k][sk] for sk in ctx[k] if sk in _L0_AGENT}
        elif k == "session":
            out[k] = {sk: ctx[k][sk] for sk in ctx[k] if sk in _L0_SESSION}
        else:
            out[k] = ctx[k]
    return out


def context_identity(ctx: dict[str, Any]) -> str:
    """``"sha256:" + hex(SHA-256(canonical_json(ctx_L01)))`` (§10.2)."""
    digest = hashlib.sha256(canonical_json(_strip_to_l01(ctx)).encode("utf-8")).hexdigest()
    return f"sha256:{digest}"
